modules/core/src/main/java/org/apache/ignite/internal/processors/cache/PartitionUpdateCounter.java

```diff
@@ -180,14 +181,24 @@ private void offer(Item item) {
 
     /**
      * Flushes pending update counters closing all possible gaps.
+     *
+     * @param onGapClose Closure to run on gap closing.
      */
-    public synchronized void finalizeUpdateCountres() {
-        Item last = queue.pollLast();
+    public synchronized void finalizeUpdateCounters(IgniteInClosure<Long> onGapClose) {
+        Item item = poll();
+
+        while (item != null) {
+            long cur = cntr.get();
 
-        if (last != null)
-            update(last.start + last.delta);
+            // Call closure for each counter between current value and next range start.
+            for (long cntr0 = cur; cntr0 < item.start; cntr0++)
+                onGapClose.apply(cntr0 + 1);
```

Comment:
Just for understanding. Why do we need to notify the listener for each counter with step equal to 1? I am thinking of possible hundreds or thousands of notifications (or even more).

Reply:
The reason ow this action is to fill backup CQ queue with the "fake" entries instead of gaps. Of cause we can think about optimization but it is not necessary here because every listener notification here is a simple put to queue.